File: frontend/lib/api/auto-fix-quota.ts
import { apiFetch } from "./client";

/**
 * Per-scan AI Auto-Fix allowance.
 *
 * Scoped to a scan, not a project: each scan is one repo at one commit, so the
 * budget refills naturally the next time that repo is scanned. `used` counts
 * distinct findings that already have a generated proposal, so regenerating or
 * revising a fix never charges twice.
 */
export type ScanAutoFixQuota = {
  scan_id: string;
  project_id: string;
  /** Global base allowance (Settings → Auto-Fix). */
  default_limit: number;
  /** Extra headroom an admin approved for this scan. */
  extra_granted: number;
  /** default_limit + extra_granted. */
  limit: number;
  used: number;
  remaining: number;
  /** Non-zero means a request is already awaiting review — don't offer another. */
  pending_request_count: number;
};

export type AutoFixQuotaRequestStatus = "pending" | "approved" | "rejected";

export type AutoFixQuotaRequest = {
  id: string;
  scan_id: string;
  project_id: string;
  project_name: string | null;
  requested_by: string;
  requested_by_email: string | null;
  requested_additional: number;
  reason: string;
  status: AutoFixQuotaRequestStatus;
  granted_additional: number | null;
  decision_note: string | null;
  decided_by: string | null;
  decided_by_email: string | null;
  decided_at: string | null;
  created_at: string;
};

export type AutoFixQuotaRequestList = {
  items: AutoFixQuotaRequest[];
  /** Full queue depth, not the filtered view — safe for a nav badge. */
  pending_count: number;
};

export function getScanAutoFixQuota(scanId: string) {
  return apiFetch<ScanAutoFixQuota>(`/scans/${scanId}/auto-fix/quota`);
}

export function listScanQuotaRequests(scanId: string) {
  return apiFetch<AutoFixQuotaRequestList>(`/scans/${scanId}/auto-fix/quota/requests`);
}

export function requestAutoFixQuota(
  scanId: string,
  body: { requested_additional: number; reason: string }
) {
  return apiFetch<AutoFixQuotaRequest>(`/scans/${scanId}/auto-fix/quota/requests`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

// --- admin ------------------------------------------------------------------

export function listAllQuotaRequests(status?: AutoFixQuotaRequestStatus) {
  const qs = status ? `?status=${status}` : "";
  return apiFetch<AutoFixQuotaRequestList>(`/admin/auto-fix-quota/requests${qs}`);
}

export function decideQuotaRequest(
  requestId: string,
  body: { approve: boolean; granted_additional?: number; decision_note?: string }
) {
  return apiFetch<AutoFixQuotaRequest>(
    `/admin/auto-fix-quota/requests/${requestId}/decide`,
    { method: "POST", body: JSON.stringify(body) }
  );
}

File: backend/app/schemas/auto_fix_quota.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.auto_fix_quota import AutoFixQuotaRequestStatus


class ScanAutoFixQuotaResponse(BaseModel):
    scan_id: str
    project_id: str
    #: Global base allowance (RemediationSettings.auto_fix_findings_per_scan).
    default_limit: int
    #: Extra headroom approved by an admin for this scan.
    extra_granted: int
    #: default_limit + extra_granted.
    limit: int
    #: Distinct findings on this scan that already have a fix proposal.
    used: int
    remaining: int
    #: Lets the UI show "request pending" instead of offering a second request.
    pending_request_count: int


class AutoFixQuotaRequestCreate(BaseModel):
    requested_additional: int = Field(ge=1, le=500)
    #: Required. An admin approving extra AI spend needs the justification.
    reason: str = Field(min_length=1, max_length=2000)


class AutoFixQuotaRequestDecision(BaseModel):
    approve: bool
    #: Omit to grant exactly what was requested; set to grant a different amount.
    granted_additional: int | None = Field(default=None, ge=1, le=500)
    decision_note: str | None = Field(default=None, max_length=2000)


class AutoFixQuotaRequestOut(BaseModel):
    id: str
    scan_id: str
    project_id: str
    #: Resolved for display so the admin table needs no extra round-trips.
    project_name: str | None = None
    requested_by: str
    requested_by_email: str | None = None
    requested_additional: int
    reason: str
    status: AutoFixQuotaRequestStatus
    granted_additional: int | None = None
    decision_note: str | None = None
    decided_by: str | None = None
    decided_by_email: str | None = None
    decided_at: datetime | None = None
    created_at: datetime


class AutoFixQuotaRequestListResponse(BaseModel):
    items: list[AutoFixQuotaRequestOut]
    pending_count: int
